var/spack/repos/builtin/packages/py-torch/package.py

```diff
@@ -104,7 +99,17 @@ class PyTorch(PythonPackage, CudaPackage, ROCmPackage):
     )
 
     conflicts("+cuda+rocm")
-    conflicts("+tensorpipe", when="+rocm", msg="TensorPipe doesn't yet support ROCm")
+    conflicts("+tensorpipe", when="^hip@:5.1")
+    conflicts(
+        "~tensorpipe",
+        when="@1.8: +distributed",
+        msg="Distributed requires TensorPipe for py-torch@1.8:",
+    )
+    conflicts(
+        "+tensorpipe",
+        when="~distributed",
+        msg="TensorPipe requires Distributed.",  # super-Oxford comma
+    )
```

Comment:
I removed the `when=` variant, since this way I get better error messages than "can't find any working combination".

You can try adding it back, but be sure that all these combinations still concretize:
- `~distributed+rocm`
- `+distributed+rocm`
- `+distributed~tensorpipe+rocm`
- `~distributed`
- `+distributed`
- `+distributed~tensorpipe`

Reply:
I would prefer keeping the variant `when` instead of add this extra conflict. It's more technically correct and matches the cmake definition.